examples/read_ogr_arrow.rs

```diff
@@ -38,50 +40,31 @@ fn run() -> gdal::errors::Result<()> {
     options.set_name_value("INCLUDE_FID", "NO")?;
 
     // Read the layer's data into our provisioned pointer
-    unsafe { layer_a.read_arrow_stream(gdal_pointer, &options).unwrap() }
+    unsafe { layer.read_arrow_stream(gdal_pointer, &options)? }
 
-    // The rest of this example is arrow2-specific.
+    // The rest of this example is specific to the `arrow` crate.
 
-    // `arrow2` has a helper class `ArrowArrayStreamReader` to assist with iterating over the raw
+    // `arrow` has a helper class `ArrowArrayStreamReader` to assist with iterating over the raw
     // batches
-    let mut arrow_stream_reader =
-        unsafe { arrow2::ffi::ArrowArrayStreamReader::try_new(output_stream).unwrap() };
+    let arrow_stream_reader = ArrowArrayStreamReader::try_new(output_stream).unwrap();
+
+    // Get the index of the geometry column
+    let geom_column_index = arrow_stream_reader
+        .schema()
+        .column_with_name("wkb_geometry")
+        .unwrap()
+        .0;
 
     // Iterate over the stream until it's finished
-    // arrow_stream_reader.next() will return None when the stream has no more data
-    while let Some(maybe_array) = unsafe { arrow_stream_reader.next() } {
+    for maybe_array in arrow_stream_reader {
         // Access the contained array
         let top_level_array = maybe_array.unwrap();
 
-        // The top-level array is a single logical "struct" array which includes all columns of the
-        // dataset inside it.
-        assert!(
-            matches!(top_level_array.data_type(), DataType::Struct(..)),
-            "Top-level arrays from OGR are expected to be of struct type"
-        );
-
-        // Downcast from the Box<dyn Array> to a concrete StructArray
-        let struct_array = top_level_array
-            .as_any()
-            .downcast_ref::<StructArray>()
-            .unwrap();
-
-        // Access the underlying column metadata and data
-        // Clones are cheap because they do not copy the underlying data
-        let (fields, columns, _validity) = struct_array.clone().into_data();
-
-        // Find the index of the geometry column
-        let geom_column_index = fields
-            .iter()
-            .position(|field| field.name == "wkb_geometry")
-            .unwrap();
-
-        // Pick that column and downcast to a BinaryArray
-        let geom_column = &columns[geom_column_index];
-        let binary_array = geom_column
-            .as_any()
-            .downcast_ref::<BinaryArray<i32>>()
-            .unwrap();
+        // Get the geometry column
+        let geom_column = top_level_array.column(geom_column_index);
+
+        // Downcast it to a `BinaryArray`
+        let binary_array = geom_column.as_any().downcast_ref::<BinaryArray>().unwrap();
```

Comment:
I don't know the exact circumstances but I thought that _sometimes_ GDAL would create an Arrow array with i64 offsets instead of i32 offsets. Maybe this is only in the rare case where the array is very large? I thought I hit a bug related to this, but it was probably in my own code.

Reply:
Gah, why is that even a thing :confused:. [I guess](https://github.com/OSGeo/gdal/blob/37651a915cdd8f1d1d1f92e0cdc9f42b12b788f5/ogr/ogrsf_frmts/generic/ogrlayerarrow.cpp#L171-L178) the offset size is part of the data type?

Anyway, we're keeping the old behavior, so I'm not sure how to improve this.

Reply:
I think as-is is fine to merge.

With Arrow, a list array has a values buffer plus a buffer of offsets. If the values buffer has more than 2^31 items (signed ints to support Java) then you can't represent those values. Usually the preferred way to fix this is to use batching, so that you never have 2^31 values in a single contiguous array, but Arrow also defines "large" nested array types, which use int64 offsets instead of int32.

I don't recall if GDAL will ever create an array with int64 offsets